csi.proto

```diff
@@ -262,6 +274,147 @@ message Volume {
+  // accepted_topologies =
+  //   {"region": {"us-west1"},
+  //    "zone": {"us-west1-b", "us-west1-c", "us-west1-d"}}
+  // Indicates the volume may be accessible from any combination of two
+  // "zones" in the "region" "us-west1": e.g. "us-west1-b" and
+  // "us-west1-c" or "us-west1-b" and "us-west1-d", etc.
+  map<string, ListOfString> accepted_topologies = 1;
+
+  // If a CO specifies multiple values for a topological segment, it may
+  // also specify one or more preferred values.  An SP SHOULD attempt
+  // to fullfil the provisioning request using the preferred value(s) in
+  // order from first to last and only use the the other
+  // accepted_topologies values if it is not possible to fulfill the
+  // CreateVolume request with the preferred value(s).
+  // Each key MUST also be present in accepted_topologies.
+  // Each value MUST be present in the list of values for the
+  // corresponding key in accepted_topologies.
+  // This field is OPTIONAL. If it is not specified, the SP MAY choose
+  // from any of the set of possible values specified in
+  // accepted_topologies.
+  //
+  // Example 1:
+  // Given a volume that is accessible from a single zone, and
+  // accepted_topologies =
+  //   {"region": {"us-west1"}, "zone": {"us-west1-b", "us-west1-c"}}
+  // preferred_topologies =
+  //   {"zone": {"us-west1-c"}}
+  // Indicates the SP SHOULD first attempt to provision the volume as
+  // accessible from "zone" "us-west1-c" in the "region" "us-west1" and
+  // fall back to "zone" "us-west1-b" in the "region" "us-west1" if that
+  // is not possible.
+  //
+  // Example 2:
+  // Given a volume that is accessible from a single zone, and
+  // accepted_topologies =
+  //   {"region": {"us-west1"},
+  //   "zone": {"us-west1-b", "us-west1-c", "us-west1-d", "us-west1-e"}}
+  // preferred_topologies =
+  //   {"zone": {"us-west1-d", "us-west1-b"}}
+  // Indicates the SP SHOULD first attempt to provision the volume as
+  // accessible from "zone" "us-west1-d" in the "region" "us-west1" and
+  // fall back to "zone" "us-west1-b" in the "region" "us-west1" if that
+  // is not possible. If that is not possible, the SP may choose between
+  // either the "zone" "us-west1-c" or "us-west1-e" in the "region"
+  // "us-west1".
+  //
+  // Example 3:
+  // Given a volume that is accessible from two zones (which may be
+  // indicated via an opaque parameter if the SP supports both single
+  // and multi-zone volumes), and
+  // accepted_topologies =
+  //   {"region": {"us-west1"},
+  //   "zone": {"us-west1-b", "us-west1-c", "us-west1-d", "us-west1-e"}}
+  // preferred_topologies =
+  //   {"zone": {"us-west1-e", "us-west1-c"}}
+  // Indicates the SP SHOULD first attempt to provision the volume as
+  // accessible from the combination of the two "zones" "us-west1-c" and
+  // "us-west1-e" in the "region" "us-west1" and fall back to other
+  // combinations if that is not possible.
+  map<string, ListOfString> preferred_topologies = 2;
```

Comment:
Also, I assume you tried to use the repeated keyword with the map value? That’s not supported?

Reply:
James, 

The reason for my suggestion is in part due to the way a custom type prevents the ability to create interface assertions on types in a language like Go without referencing the language bindings.

Reply:
That’s fine. Has anyone tried a repeated keyword inside the map value?

Reply:
Nevermind.

<img width="1195" alt="screencapture-2018 02 13-15 31 07" src="https://user-images.githubusercontent.com/101085/36174775-f7bce214-10d2-11e8-9074-9be03ed3ef2c.png">

I still believe using an existing type rather than defining a new one is preferred for downstream language bindings.